ParallelPrimitives/RadixSortKernels.h

```diff
@@ -521,278 +514,708 @@ extern "C" __global__ void SortSinglePassKernel( int* gSrcKey, int* gDstKey, int
 
 	__syncthreads();
 
-	// Do parallel exclusive scan on the LDS
+	prefixSumExclusive<256>( 0, smem );
 
-	int currentSegmentSum = ldsScanExclusive( blockBuffer, SCAN_WG_SIZE );
-
-	__syncthreads();
-
-	// Sync all the Workgroups to calculate the global offset.
-
-	__shared__ int currentGlobalOffset;
-	WorkgroupSync( threadIdx.x, blockIdx.x, currentSegmentSum, &currentGlobalOffset, gPartialSum, gIsReady );
-
-	// Write back the result.
-
-	gHistogram[blockIdx.x * blockDim.x + threadIdx.x] = blockBuffer[threadIdx.x] + currentGlobalOffset;
+	gpSumBuffer[blockIdx.x * 256 + threadIdx.x] = smem[threadIdx.x];
 }
 
-template<bool KEY_VALUE_PAIR>
-__device__ void SortImpl( int* gSrcKey, int* gSrcVal, int* gDstKey, int* gDstVal, int* gHistogram, int numberOfInputs, int gNItemsPerWG, const int START_BIT, const int N_WGS_EXECUTED )
+__device__ __forceinline__ void onesweep_reorder( RADIX_SORT_KEY_TYPE* inputKeys, RADIX_SORT_KEY_TYPE* outputKeys, RADIX_SORT_VALUE_TYPE* inputValues, RADIX_SORT_VALUE_TYPE* outputValues, bool keyPair, uint32_t numberOfInputs, uint32_t* gpSumBuffer,
+												  volatile uint64_t* lookBackBuffer, uint32_t startBits, uint32_t iteration )
 {
-	const int startOffset = blockIdx.x * gNItemsPerWG;
-	const int nItemInBlock = ( startOffset + gNItemsPerWG > numberOfInputs ) ? numberOfInputs - startOffset : gNItemsPerWG;
-
 	struct ElementLocation
 	{
-		u32 localSrcIndex : 12;
-		u32 localOffset : 12;
-		u32 bucket : 8;
+		uint32_t localSrcIndex : 12;
+		uint32_t localOffset : 12;
+		uint32_t bucket : 8;
 	};
 
-	__shared__ u32 globalOffset[BIN_SIZE];
-	__shared__ u32 localPrefixSum[BIN_SIZE];
-	__shared__ u32 counters[BIN_SIZE];
-	__shared__ u32 matchMasks[SORT_NUM_WARPS_PER_BLOCK][BIN_SIZE];
-	__shared__ u8 elementBuckets[SORT_SUBBLOCK_SIZE];
-	__shared__ ElementLocation elementLocations[SORT_SUBBLOCK_SIZE];
+	__shared__ uint32_t pSum[256];
+	__shared__ uint32_t localPrefixSum[256];
+	__shared__ uint32_t counters[256];
+	__shared__ ElementLocation elementLocations[RADIX_SORT_BLOCK_SIZE];
+	__shared__ uint8_t elementBuckets[RADIX_SORT_BLOCK_SIZE];
+	__shared__ uint32_t matchMasks[REORDER_NUMBER_OF_WARPS][256];
 
-	for( int i = threadIdx.x; i < BIN_SIZE; i += SORT_WG_SIZE )
-	{
-		// Note: The size of gHistogram is always BIN_SIZE * N_WGS_EXECUTED
-		globalOffset[i] = gHistogram[i * N_WGS_EXECUTED + blockIdx.x];
-	}
+	uint32_t bitLocation = startBits + 8 * iteration;
+	uint32_t blockIndex = blockIdx.x;
+	uint32_t numberOfBlocks = div_round_up( numberOfInputs, RADIX_SORT_BLOCK_SIZE );
 
-	for( int w = 0; w < SORT_NUM_WARPS_PER_BLOCK; ++w )
+	clearShared<256, REORDER_NUMBER_OF_THREADS_PER_BLOCK, uint32_t>( localPrefixSum, 0 );
+	clearShared<256, REORDER_NUMBER_OF_THREADS_PER_BLOCK, uint32_t>( counters, 0 );
+
+	for( int w = 0; w < REORDER_NUMBER_OF_WARPS; w++ )
 	{
-		for( int i = threadIdx.x; i < BIN_SIZE; i += SORT_WG_SIZE )
+		for( int i = 0; i < 256; i += REORDER_NUMBER_OF_THREADS_PER_BLOCK )
```

Comment:
You can start from `threadIdx.x` instead of 0, but it's a personal preference .

Reply:
Maybe they should be consistent at least, I unified to threadIdx.x begining